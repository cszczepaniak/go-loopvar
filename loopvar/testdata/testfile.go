package testdata

import "fmt"

func slice() {
	for i, v := range []int{1, 2, 3} {
		i := i // want "found unnecessary loop variable capture"
		v := v // want "found unnecessary loop variable capture"

		fmt.Println(i, v)
	}
}

func statementsInBetween() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"

		for range 10 {
		}

		go func() {}()
		val := v // want "found unnecessary loop variable capture"

		fmt.Println(index, val)
	}
}

func multiAssign() {
	for i, v := range []int{1, 2, 3} {
		idx, v := i, v // want "found unnecessary loop variable capture"

		fmt.Println(idx, v)
	}
}

func trickyMultiAssign0() {
	for i, v := range []int{1, 2, 3} {
		_, anotherVar, val := i, 123, v // want "found unnecessary loop variable capture"

		fmt.Println(i, val, anotherVar)
	}
}

func trickyMultiAssign1() {
	for i, v := range []int{1, 2, 3} {
		_, index, anotherVar, val := 456, i, 123, v // want "found unnecessary loop variable capture"

		fmt.Println(index, val, anotherVar)
	}
}

func trickyMultiAssign2() {
	for _, v := range []int{1, 2, 3} {
		_, val := func() int { return 1 }(), v // want "found unnecessary loop variable capture"

		fmt.Println(val)
	}
}

func oneVariableIsIncrementedLaterOneIsNot() {
	for i, v := range []int{1, 2, 3} {
		incrementing := i
		variable := v // want "found unnecessary loop variable capture"

		incrementing++

		fmt.Println(incrementing, variable)
	}
}

func variableIsIncrementedLater() {
	for i, v := range []int{1, 2, 3} {
		incrementing, variable := i, v // want "found unnecessary loop variable capture"

		incrementing++

		fmt.Println(incrementing, variable)
	}
}

func variableIsDecrementedLater() {
	for i, v := range []int{1, 2, 3} {
		decrementing, variable := i, v // want "found unnecessary loop variable capture"

		decrementing--

		fmt.Println(decrementing, variable)
	}
}

func variableIsPlusAssigned() {
	for i, v := range []int{1, 2, 3} {
		incrementing, variable := i, v // want "found unnecessary loop variable capture"

		incrementing += 123

		fmt.Println(incrementing, variable)
	}
}

func variableIsMinusAssigned() {
	for i, v := range []int{1, 2, 3} {
		decrementing, variable := i, v // want "found unnecessary loop variable capture"

		decrementing -= 123

		fmt.Println(decrementing, variable)
	}
}

func variableIsMultiplyAssigned() {
	for i, v := range []int{1, 2, 3} {
		index, variable := i, v // want "found unnecessary loop variable capture"

		index *= 123

		fmt.Println(index, variable)
	}
}

func variableIsDivideAssigned() {
	for i, v := range []int{1, 2, 3} {
		index, variable := i, v // want "found unnecessary loop variable capture"

		index /= 123

		fmt.Println(index, variable)
	}
}

func variableIsModuloAssigned() {
	for i, v := range []int{1, 2, 3} {
		index, variable := i, v // want "found unnecessary loop variable capture"

		index %= 123

		fmt.Println(index, variable)
	}
}

func variableIsAndAssigned() {
	for i, v := range []int{1, 2, 3} {
		index, variable := i, v // want "found unnecessary loop variable capture"

		index &= 123

		fmt.Println(index, variable)
	}
}

func variableIsOrAssigned() {
	for i, v := range []int{1, 2, 3} {
		index, variable := i, v // want "found unnecessary loop variable capture"

		index |= 123

		fmt.Println(index, variable)
	}
}

func variableIsXorAssigned() {
	for i, v := range []int{1, 2, 3} {
		index, variable := i, v // want "found unnecessary loop variable capture"

		index ^= 123

		fmt.Println(index, variable)
	}
}

func variableIsShiftLeftAssigned() {
	for i, v := range []int{1, 2, 3} {
		index, variable := i, v // want "found unnecessary loop variable capture"

		index <<= 123

		fmt.Println(index, variable)
	}
}

func variableIsShiftRightAssigned() {
	for i, v := range []int{1, 2, 3} {
		index, variable := i, v // want "found unnecessary loop variable capture"

		index >>= 123

		fmt.Println(index, variable)
	}
}

func variableIsAndNotAssigned() {
	for i, v := range []int{1, 2, 3} {
		index, variable := i, v // want "found unnecessary loop variable capture"

		index &^= 123

		fmt.Println(index, variable)
	}
}

func variableIsSimplyAssigned() {
	for i, v := range []int{1, 2, 3} {
		assigned, variable := i, v // want "found unnecessary loop variable capture"

		assigned = 123

		fmt.Println(assigned, variable)
	}
}

func wasABugBeforeGo122() {
	for i, v := range []int{1, 2, 3} {
		go func() {
			fmt.Println(i, v)
		}()

		idx := i // want "found unnecessary loop variable capture"
		val := v // want "found unnecessary loop variable capture"

		fmt.Println(idx, val)
	}
}

func muchWhitespace() {
	for i, v := range []int{1, 2, 3} {

		i := i // want "found unnecessary loop variable capture"

		v := v // want "found unnecessary loop variable capture"

		fmt.Println(i, v)
	}
}

func aMap() {
	for k, v := range map[string]int{} {
		k := k // want "found unnecessary loop variable capture"
		v := v // want "found unnecessary loop variable capture"

		fmt.Println(k, v)
	}
}

func rename() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"
		val := v   // want "found unnecessary loop variable capture"

		fmt.Println(index, val)
	}
}

func address() {
	vals := []*int{}
	for _, v := range []int{1, 2, 3} {
		v := v // want "found unnecessary loop variable capture"
		vals = append(vals, &v)
	}
	_ = vals
}

func addressRename() {
	vals := []*int{}
	for _, v := range []int{1, 2, 3} {
		myValue := v // want "found unnecessary loop variable capture"
		vals = append(vals, &myValue)
	}
	_ = vals
}

func renameInGoroutine() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"
		val := v   // want "found unnecessary loop variable capture"

		go func() {
			fmt.Println(index, val)
		}()
	}
}

func kitchenSink() {
	for i1, v1 := range []int{1, 2, 3} {
		i1 := i1   // want "found unnecessary loop variable capture"
		val1 := v1 // want "found unnecessary loop variable capture"

		for i2, v2 := range []string{"foo", "bar"} {
			// I'm not sure how we should handle such captures... I don't expect they'll be common,
			// though.
			varA := i1
			anothaOne := v2 // want "found unnecessary loop variable capture"

			go func() {
				for i3, v3 := range map[int]int{} {
					my3 := i3 // want "found unnecessary loop variable capture"
					v3 := v3  // want "found unnecessary loop variable capture"

					fmt.Println(
						i2,
						i1,
						val1,
						v1,
						my3,
						v3*(my3+i1-val1+v3),
						anothaOne,
						varA,
						v2,
					)
				}
			}()
		}
	}
}
