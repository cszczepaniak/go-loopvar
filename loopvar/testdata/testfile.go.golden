package testdata

import "fmt"

func slice() {
	for i, v := range []int{1, 2, 3} {
		fmt.Println(i, v)
	}
}

func statementsInBetween() {
	for i, v := range []int{1, 2, 3} {
		for range 10 {
		}

		go func() {}()
		fmt.Println(i, v)
	}
}

func multiAssign() {
	for i, v := range []int{1, 2, 3} {
		fmt.Println(i, v)
	}
}

func trickyMultiAssign0() {
	for i, v := range []int{1, 2, 3} {
		anotherVar := 123 // want "found unnecessary loop variable capture"

		fmt.Println(i, v, anotherVar)
	}
}

func trickyMultiAssign1() {
	for i, v := range []int{1, 2, 3} {
		_, anotherVar := 456, 123 // want "found unnecessary loop variable capture"

		fmt.Println(i, v, anotherVar)
	}
}

func trickyMultiAssign2() {
	for _, v := range []int{1, 2, 3} {
		_ = func() int {
			return 1
		}() // want "found unnecessary loop variable capture"

		fmt.Println(v)
	}
}

func variableIsIncrementedLater() {
	for i, v := range []int{1, 2, 3} {
		incrementing := i // want "found unnecessary loop variable capture"

		incrementing++

		fmt.Println(incrementing, v)
	}
}

func variableIsDecrementedLater() {
	for i, v := range []int{1, 2, 3} {
		decrementing := i // want "found unnecessary loop variable capture"

		decrementing--

		fmt.Println(decrementing, v)
	}
}

func variableIsPlusAssigned() {
	for i, v := range []int{1, 2, 3} {
		incrementing := i // want "found unnecessary loop variable capture"

		incrementing += 123

		fmt.Println(incrementing, v)
	}
}

func variableIsMinusAssigned() {
	for i, v := range []int{1, 2, 3} {
		decrementing := i // want "found unnecessary loop variable capture"

		decrementing -= 123

		fmt.Println(decrementing, v)
	}
}

func variableIsMultiplyAssigned() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"

		index *= 123

		fmt.Println(index, v)
	}
}

func variableIsDivideAssigned() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"

		index /= 123

		fmt.Println(index, v)
	}
}

func variableIsModuloAssigned() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"

		index %= 123

		fmt.Println(index, v)
	}
}

func variableIsAndAssigned() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"

		index &= 123

		fmt.Println(index, v)
	}
}

func variableIsOrAssigned() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"

		index |= 123

		fmt.Println(index, v)
	}
}

func variableIsXorAssigned() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"

		index ^= 123

		fmt.Println(index, v)
	}
}

func variableIsShiftLeftAssigned() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"

		index <<= 123

		fmt.Println(index, v)
	}
}

func variableIsShiftRightAssigned() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"

		index >>= 123

		fmt.Println(index, v)
	}
}

func variableIsAndNotAssigned() {
	for i, v := range []int{1, 2, 3} {
		index := i // want "found unnecessary loop variable capture"

		index &^= 123

		fmt.Println(index, v)
	}
}

func variableIsSimplyAssigned() {
	for i, v := range []int{1, 2, 3} {
		assigned := i // want "found unnecessary loop variable capture"

		assigned = 123

		fmt.Println(assigned, v)
	}
}

func wasABugBeforeGo122() {
	for i, v := range []int{1, 2, 3} {
		go func() {
			fmt.Println(i, v)
		}()

		fmt.Println(i, v)
	}
}

func muchWhitespace() {
	for i, v := range []int{1, 2, 3} {

		fmt.Println(i, v)
	}
}

func aMap() {
	for k, v := range map[string]int{} {
		fmt.Println(k, v)
	}
}

func rename() {
	for i, v := range []int{1, 2, 3} {
		fmt.Println(i, v)
	}
}

func address() {
	vals := []*int{}
	for _, v := range []int{1, 2, 3} {
		vals = append(vals, &v)
	}
	_ = vals
}

func addressRename() {
	vals := []*int{}
	for _, v := range []int{1, 2, 3} {
		vals = append(vals, &v)
	}
	_ = vals
}

func renameInGoroutine() {
	for i, v := range []int{1, 2, 3} {
		go func() {
			fmt.Println(i, v)
		}()
	}
}

func kitchenSink() {
	for i1, v1 := range []int{1, 2, 3} {
		for i2, v2 := range []string{"foo", "bar"} {
			// I'm not sure how we should handle such captures... I don't expect they'll be common,
			// though.
			varA := i1
			go func() {
				for i3, v3 := range map[int]int{} {
					fmt.Println(
						i2,
						i1,
						v1,
						v1,
						i3,
						v3*(i3+i1-v1+v3),
						v2,
						varA,
						v2,
					)
				}
			}()
		}
	}
}
